src/rabbit_connection_tracking.erl

```diff
@@ -0,0 +1,345 @@
+                      mnesia:delete({TableName, ConnId})
+              end
+      end).
+
+
+-spec list() -> [rabbit_types:tracked_connection()].
+
+list() ->
+  Chunks = lists:map(
+    fun (Node) ->
+      Tab = tracked_connection_table_name_for(Node),
+      mnesia:dirty_match_object(Tab, #tracked_connection{_ = '_'})
+    end, rabbit_mnesia:cluster_nodes(running)),
+  lists:foldl(fun(Chunk, Acc) -> Acc ++ Chunk end, [], Chunks).
+
+
+-spec list(rabbit_types:vhost()) -> [rabbit_types:tracked_connection()].
+
+list(VHost) ->
+  Chunks = lists:map(
+    fun (Node) ->
+      Tab = tracked_connection_table_name_for(Node),
+      mnesia:dirty_match_object(Tab, #tracked_connection{vhost = VHost, _ = '_'})
+    end, rabbit_mnesia:cluster_nodes(running)),
+  lists:foldl(fun(Chunk, Acc) -> Acc ++ Chunk end, [], Chunks).
+
+
+-spec list_on_node(node()) -> [rabbit_types:tracked_connection()].
+
+list_on_node(Node) ->
+  try mnesia:dirty_match_object(
+    tracked_connection_table_name_for(Node),
+    #tracked_connection{_ = '_'})
+  catch exit:{aborted, {no_exists, _}} -> []
+  end.
+
+-spec is_over_connection_limit(rabbit_types:vhost()) -> {true, non_neg_integer()} | false.
+
+is_over_connection_limit(VirtualHost) ->
+    case rabbit_vhost_limit:connection_limit(VirtualHost) of
+        %% no limit configured
+        undefined                                            -> false;
+        %% with limit = 0, no connections are allowed
+        {ok, 0}                                              -> {true, 0};
+        {ok, Limit} when is_integer(Limit) andalso Limit > 0 ->
+          ConnectionCount = count_connections_in(VirtualHost),
+          case ConnectionCount >= Limit of
+            false -> false;
+            true  -> {true, Limit}
+          end;
+        %% any negative value means "no limit". Note that parameter validation
+        %% will replace negative integers with 'undefined', so this is to be
+        %% explicit and extra defensive
+        {ok, Limit} when is_integer(Limit) andalso Limit < 0 -> false;
+        %% ignore non-integer limits
+        {ok, _Limit}                                         -> false
+    end.
+
+
+-spec count_connections_in(rabbit_types:vhost()) -> non_neg_integer().
+
+count_connections_in(VirtualHost) ->
+    try
+      Ns = lists:map(
+        fun (Node) ->
+          Tab = tracked_connection_per_vhost_table_name_for(Node),
+          try
+            case mnesia:transaction(
+               fun() ->
+                 case mnesia:dirty_read({Tab, VirtualHost}) of
+                      []    -> 0;
+                      [Val] -> Val#tracked_connection_per_vhost.connection_count
+                 end
+               end) of
+                 {atomic,  Val}     -> Val;
+                 {aborted, _Reason} -> 0
+            end
+          catch _ -> 0
+          end
+        end, rabbit_mnesia:cluster_nodes(running)),
+      lists:foldl(fun(X, Acc) -> Acc + X end, 0, Ns)
```

Comment:
Sorry but you are worrying about the wrong thing here. What really needs to happen is doing a parallel map but for small clusters it will likely be less efficient so I decided that it can be done at a later point.

It's not common to see clusters with more than maybe 7-9 nodes and it's unlikely to be a major contributor to how many connections/second can be accepted. Plus if you don't have any limits configured, none of this code will be executed.